### DISCUSSION TITLE
[SOLVED] FF108 : Extensions' dynamic toolbar icons' replacement no longer work.

### DISCUSSION THREAD

**Discussion by ghost** [Aris-t2/CustomCSSforFx#532] [Q&A]
Hi,

I have a problem **_not related_** to CustomCSSforFX, I've searched unsuccessfully on [Reddit - Firefox](https://www.reddit.com/r/firefox/) and [Reddit - Firefox CS](https://www.reddit.com/r/FirefoxCSS/ )S. I'd really appreciate your help.

This is specific to latest Firefox 108.0

I replace several Firefox extensions' .toolbarbuttons icons. I face an issue when replacing a dynamic icon with a _style_ condition, i.e:

```
/* uBlock Origin default toolbarbutton-icon : STILL WORKS */
#ublock0_raymondhill_net-browser-action .toolbarbutton-icon {list-style-image: url(images/uBlock.png) !important;}
/* uBlock Origin off toolbarbutton-icon : NO LONGER WORKS */
#ublock0_raymondhill_net-browser-action[style*="icon_16-off"] .toolbarbutton-icon {list-style-image: url(images/uBlock-OFF.png) !important;}
```

I cannot understand why the [style*="XXX"] no longer works with Firefox 108.0
I'm aware FF108 has brought its lot of CSS issues, but this issue this time is really problematic.

Many thanks for any advice, explanation, solution :=)

**Comment by ghost** [comment]
I finally found the explanation: [Firefox 108 broke all my toolbar and URL-bar icon tweaks : FirefoxCSS](https://teddit.pussthecat.org/r/FirefoxCSS/comments/zmjxkk/firefox_108_broke_all_my_toolbar_and_urlbar_icon/j0bgfoj/#c)

> Extensions buttons in the toolbar are now wrapped into a <toolbaritem> element - the selectors you are using match that toolbaritem, but if you want to change the icon used then you need to target the actual button within that wrapper.

My above example:

```
#ublock0_raymondhill_net-browser-action .toolbarbutton-icon {list-style-image: url(images/uBlock.png) !important;}
#ublock0_raymondhill_net-browser-action[style*="icon_16-off"] .toolbarbutton-icon {list-style-image: url(images/uBlock-OFF.png) !important;}
```
BECOMES
```
#ublock0_raymondhill_net-BAP {list-style-image: url(images/uBlock.png) !important;}
#ublock0_raymondhill_net-BAP[style*="icon_16-off"] {list-style-image: url(images/uBlock-OFF.png) !important;}
```
**It works!**

**Comment by Speravir** [comment]
@Cade66 See [Add distinction between active and inactive state, again · uBlockOrigin/uBlock-issues](https://github.com/uBlockOrigin/uBlock-issues/discussions/2409).

But I am surprised that using the style attribute works because I tried exactly this method for another add-on in the past and it didn’t work then, so I did not check again this time. Bad luck.

**Comment by ghost** [comment]
@Speravir indeed what I mentioned above doesn't always work as described by the source i mentioned above:

> Extensions buttons in the toolbar are now wrapped into a element - the selectors you are using match that toolbaritem, but if you want to change the icon used then you need to target the actual button within that wrapper.

The extension's ID displayed in `about:support` or in `about:debugging#/runtime/this-firefox` is the reference you'll need to find it in your  **_prefs.js_** : 

Example1- uBlock Origin
Extension ID: uBlock0@raymondhill.net
In prefs.js: ublock0_raymondhill_net-browser-action 

Example2- Stylus
Extension ID: {7a7a4a92-a2a0-41d1-9fd7-1e92480d612d}
In prefs.js: _7a7a4a92-a2a0-41d1-9fd7-1e92480d612d_-browser-action

It is always the extension's layout followed by `-browser-action` in the prefs.js file which must be considered.

Hence for Stylus we'd have i.e.

```
/* Stylus */
#_7a7a4a92-a2a0-41d1-9fd7-1e92480d612d_-BAP {list-style-image: url(images/Stylus.png) !important;}
```

BUT NOT
```
/* Stylus */
{7a7a4a92-a2a0-41d1-9fd7-1e92480d612d}-BAP {list-style-image: url(images/Stylus.png) !important;}
```

In practice:

1- Search for the extension ID in `about:support` or in `about:debugging#/runtime/this-firefox`
2- Copy that ID, remove brackets if any, remove `@` and anything after `@` and copy that.
3- Open your prefs.js file, search for the modified ID you've copied followed by -browser-action
4- Copy that from the prefs.js, i.e.` _7a7a4a92-a2a0-41d1-9fd7-1e92480d612d_-browser-action` AND replace `-browser-action` by `-BAP` then use that in your userChrome.css file: **DO NOT EDIT/MODIFY YOUR prefs.js file!**

That's all.

If you wish to modify a dynamic icon (one which changes accordingly to the extension's status) then the process is the same except that you'll add a style condition, i.e.:

`#ublock0_raymondhill_net-BAP[style*="icon_16-off"] {list-style-image: url(images/uBlock-OFF.png) !important;}`

Here, `[style*="icon_16-off"]` refers to the icon used by the extension according to its status. To find that icon you'll have to open the extension's xpi and search for its icons, figure out which one corresponds to the one you wish to modify and add it to `style*="[ICON=NAME"]`
`style=` requires the exact icon name, `style*=` means the name needs only to start with (or include, I forgot) `[ICON=NAME"]`

Hope that helps. I'm no expert so my way of proceeding and of explaining is obviously not at all pro, but the logic is valid and the result is guaranteed to work

Don't hesitate to ask me anything I wouldn't have sufficiently well explained.

**Reply by Speravir** [reply]
_I_ know this, as you could see in the other discussion I’ve linked above. It may help other people, though. I just was surprised that `[style*="icon_16-off"]` works.

**Reply by ghost** [reply]
I detailed my lengthy reply because as I understood it you were facing issues with the style attribute and, in doubt, included the whole scheme:

    But I am surprised that using the style attribute works because I tried exactly this method for another add-on in the past and it didn’t work then, so I did not check again this time. Bad luck.

If all runs correctly then great:

    I know this, as you could see in the other discussion I’ve linked above. It may help other people, though. I just was surprised that [style*="icon_16-off"] works.

Concerning the label attribute applied to uBlock Origin, I cannot provide an experience with it. Previously to the label attribute I've used the tooltip attribute until it was removed by uBlock Origin, which is why, then, I searched for a work-around and discovered the style attribute on a Reddit post.

/* uBlock Origin */
#ublock0_raymondhill_net-browser-action .toolbarbutton-icon {list-style-image: url(UserData/ublock.png) !important;}
#ublock0_raymondhill_net-browser-action[tooltiptext="uBlock Origin (off)"] .toolbarbutton-icon {opacity: 0.40 !important;}

Whatever, no doubt Firefox 108.0 had brought its lot of CSS modifications which bother more than one user committed to interface tweaks.
